streaming/src/main/scala/org/apache/gearpump/streaming/appmaster/ClockService.scala

```diff
@@ -75,10 +82,12 @@ class ClockService(dag : DAG, store: AppDataStore) extends Actor with Stash {
       processorIdWithDescription =>
         val (processorId, description) = processorIdWithDescription
         val taskClocks = new Array[TimeStamp](description.parallelism).map(_ => startClock)
-        val processorClock = new ProcessorClock(processorId, startClock, taskClocks)
-        processorClocks.add(processorClock)
+        val processorClock = new ProcessorClock(taskClocks)
         processorClockLookup.put(processorId, processorClock)
     }
+
+    this.processorIdToLevel = dag.graph.topologicalOrderIterator.zipWithIndex.toMap
```

Comment:
We will add a virtual link between these two processors, so that the level can be ordered.